packages/pluggableWidgets/carousel-web/src/components/Carousel.tsx

```diff
@@ -0,0 +1,45 @@
+import { createElement, ReactNode, ReactElement } from "react";
+import { ObjectItem } from "mendix";
+import Swiper, { SwiperOptions, Navigation, Pagination, EffectFade, Autoplay } from "swiper";
+import { Swiper as SwiperReact, SwiperSlide } from "swiper/react";
+
+export interface CarouselProps<T extends ObjectItem> {
+    pagination: boolean;
+    loop: boolean;
+    tabIndex?: number | undefined;
+    animation?: boolean;
+    autoplay?: boolean;
+    delay?: number;
+    navigation: boolean;
+    onClick: () => void;
+    items: T[];
+    itemRenderer: (renderWrapper: (children: ReactNode) => ReactElement, item: T) => ReactNode;
+}
+
+export function Carousel<T extends ObjectItem>(props: CarouselProps<T>): ReactElement {
+    const { items, pagination, loop, tabIndex, animation, autoplay, delay, navigation, itemRenderer, onClick } = props;
+    Swiper.use([Navigation, Pagination, EffectFade, Autoplay]);
+    const options: SwiperOptions = {
+        slidesPerView: 1,
+        centeredSlides: true,
+        loop,
+        navigation,
+        autoplay: autoplay ? { delay, stopOnLastSlide: true } : false,
+        pagination: pagination ? { type: "bullets", clickable: true } : false,
+        ...(animation && {
+            effect: "fade",
+            fadeEffect: { crossFade: true }
+        })
+    };
+    return (
+        <div className={"widget-carousel"} tabIndex={tabIndex}>
+            <SwiperReact {...options} onClick={onClick}>
+                {items.length > 0 &&
+                    itemRenderer &&
+                    items?.map((item: T, key: number) =>
+                        itemRenderer((children: ReactNode) => <SwiperSlide key={key}>{children}</SwiperSlide>, item)
```

Comment:
Let's not use the index as a key for this 😅  That's 100% gonna lead to bugs. `item.id` would be better

Reply:
which kind of bugs?

Reply:
https://reactjs.org/docs/lists-and-keys.html#keys

Reply:
Yes, but we don't use any pagination, filtering, sorting and other type of functions here.
And as it says in documentation.
`We don’t recommend using indexes for keys if the order of items may change.`

Reply:
I think we should use the GUID instead, which comes from `${item.id} ?? item_${index}` if I am not wrong, same as Data grid, otherwise when the object updates is removed from the datasource it might still render the old data because the index will still be same.

Reply:
@raffiar please check above comment

Reply:
Also users in mendix can have data set up for this widget which can be changed based on something else on the page or another widget, which will cause the exact issues we discussed

Reply:
Now it makes sense